### PR TITLE
add extended error handling

### PR DESCRIPTION
**Pull request by Stebalien**
fixes #19

### CODE REVIEW

File: response.go [hunk 1]
```diff
@@ -13,6 +13,48 @@ import (
 	"os"
 )
 
+// Error codes adapted from go-ipfs-cmds. We should find a better solution.
+
+// ErrorType signfies a category of errors
+type ErrorType uint
```
Comment (Stebalien) [inline response.go]: @magik6k ideas where to put this?

**Comment by Stebalien** [comment]
cc @MichaelMure.

File: response.go (hunk 1 above)
Comment (magik6k) [reply response.go]: I'm assuming that this is mirrored from go-ipfs-cmds - if that's the case I'd just import cmds here as it really what is implemented here

**Comment by MichaelMure** [comment]
lgtm

It might be interesting for debuggability to also have access to the HTTP status code, but at least this allow to implement backoff and it provide an easy and consistent semantic to handle errors.

**Comment by Stebalien** [comment]
> It might be interesting for debuggability to also have access to the HTTP status code, but at least this allow to implement backoff and it provide an easy and consistent semantic to handle errors.

The concern is that this interface tries to abstract over the API transport. I'd like to avoid leaking anything transport specific into it.

File: response.go (hunk 1 above)
Comment (Stebalien) [reply response.go]: Done. I'd rather not depend on that library in the long run but that's probably the best solution for now.

**Comment by Stebalien** [comment]
> The concern is that this interface tries to abstract over the API transport. I'd like to avoid leaking anything transport specific into it.

Hm. Actually, it doesn't. This is an internal error. We could have a separate HttpStatus code.

**Comment by Stebalien** [comment]
Note: this current change is slightly API breaking as it removes the `Command` field. Thoughts?

**Comment by magik6k** [comment]
> Note: this current change is slightly API breaking as it removes the Command field. Thoughts?

I'm fine with this

**Comment by Stebalien** [comment]
> It might be interesting for debuggability to also have access to the HTTP status code, but at least this allow to implement backoff and it provide an easy and consistent semantic to handle errors.

> Hm. Actually, it doesn't. This is an internal error. We could have a separate HttpStatus code.

(punting)

**Pull request merged**